### PR TITLE
Teko: Use implicit operator for Block Jacobi

### PR DESCRIPTION
**Pull request by MalachiTimothyPhillips**
This is required as ifpack2::relaxation preconditioners do not support beta != 0.

### CODE REVIEW

**Comment by MalachiTimothyPhillips** [comment]
@cgcgcg

File: packages/teko/src/Teko_BlockDiagonalInverseOp.cpp [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+                                           const double alpha, const double beta) const {
+  // call the no tranpose version
+  implicitApply(Thyra::NOTRANS, src, dst, alpha, beta);
+}
+
+/** @brief Perform a matrix vector multiply with this operator.
+ *
+ * The <code>apply</code> function takes one vector as input
+ * and applies the inverse \f$ LDU \f$ decomposition. The result
+ * is returned in \f$y\f$. If this operator is reprsented as \f$M\f$ then
```
Comment (cgcgcg) [inline packages/teko/src/Teko_BlockDiagonalInverseOp.cpp:50]: reprsented -> represented

I am a bit confused about the "ignoring conjugation!" part. Isn't the method using `M_trans`?

I think the docstrings should only be in the .hpp.

File: packages/teko/src/Teko_BlockDiagonalInverseOp.hpp [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+      productDomain_;  ///< Domain vector space.
+
+  // scratch space...so we don't have to reallocate
+  mutable BlockedMultiVector srcScrap_;
+  mutable BlockedMultiVector dstScrap_;
+  mutable bool allocated = false;
+
+ private:
+  // hide me!
+  BlockDiagonalInverseOp();
```
Comment (cgcgcg) [inline packages/teko/src/Teko_BlockDiagonalInverseOp.hpp:99]: Should the just be marked as `BlockDiagonalInverseOp() = delete;`?

File: packages/teko/src/Teko_BlockDiagonalInverseOp.cpp (hunk 1 above)
Comment (MalachiTimothyPhillips) [reply packages/teko/src/Teko_BlockDiagonalInverseOp.cpp]: I'll be honest -- most of this is copied over from https://github.com/trilinos/Trilinos/blob/26e307fb0fba3800525aef0870cea2a6dc2fd625/packages/teko/src/Teko_BlockUpperTriInverseOp.hpp#L52! 😅

**Review by cgcgcg — APPROVED** [review]

**Comment by MalachiTimothyPhillips** [comment]
> Status Flag 'Pull Request AutoTester' - Jenkins Testing: 1 or more Jobs FAILED
> 
> Note: Testing will normally be attempted again in approx. 2 Hrs 30 Mins. If a change to the PR source branch occurs, the testing will be attempted again on next available autotester run.
> Pull Request Auto Testing has FAILED (click to expand)
> 
> [CDash Test Results for PR# 13322.](https://trilinos-cdash.sandia.gov/index.php?project=Trilinos&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=63&value1=PR-13322-test&field2=buildstarttime&compare2=84&value2=NOW)
> 
> [Wiki: How to Reproduce PR Testing Builds and Errors.](https://github.com/trilinos/Trilinos/wiki/Reproducing-PR-Testing-Errors)

@cgcgcg Any idea what's going on here? Half of the tests are running the wrong revision and a bunch fail to install yet no compilation errors are listed. The previous test passed at https://github.com/trilinos/Trilinos/pull/13322#issuecomment-2270309381, and the only changes made since then are the relatively minor ones suggested in code review from https://github.com/trilinos/Trilinos/pull/13322/commits/a03f7f9537432749cb24d91641a5ae3a63e28060.

**Comment by cgcgcg** [comment]
@trilinos/framework @sebrowne

**Comment by sebrowne** [comment]
The three GCC builds are failing, which almost certainly indicates a warning-as-error issue.  I cannot explain why CDash does not display the literal error messages, but does show that an error was detected.  I can refer this to Kitware and see if they have any ideas.

**Comment by sebrowne** [comment]
@malphil yes that is technically possible depending on when the werror flag setting went in (I didn't check the dates/times)

**Review by cgcgcg — APPROVED** [review]

**Pull request merged**